[stdlib] Improve documentation for the various "String from C-string" initializers

glessard opened the pull request:
@natecook1000 pointed out missing documentation in https://github.com/apple/swift/pull/68423#discussion_r1447992994

This PR adds doc-comments to the `Array`-taking overloads of the C-string conversion initializers. In the cases where there is a more appropriate Swift spelling, they are now deprecated in Swift 6, with a message pointing to the better alternative both in the deprecation declaration and in the doc-comment.

Other minor wording improvements included.

Addresses rdar://121395821

glessard committed:
[stdlib] fix preconditionFailure message

glessard committed:
[stdlib] mark deprecated in swift 6, point to better API

glessard committed:
[stdlib] remove useless annotations

glessard committed:
[stdlib] improve phrasing in some doc-comments

glessard committed:
[stdlib] document some deprecated initializers

glessard commented on stdlib/public/core/CString.swift:
```diff
@@ -400,13 +451,16 @@ extension String {
-  ///     before the first zero code unit.
+  ///   - nullTerminatedCodeUnits: A pointer to a null-terminated sequence of
+  ///     code units encoded in `sourceEncoding`.
   ///   - sourceEncoding: The encoding in which the code units should be
```
Should this parameter label be changed to `encoding` as in all the other cases?

natecook1000 replied:
Up to you! The only effect of that name change is in the documentation, so it's fine to update it. And while consistency is good, it's unlikely that people will be comparing these parameter names directly, since they don't appear on the method listing page.

glessard replied:
I think I'll do that label change. Thanks for the soundness check.

glessard commented:
@swift-ci please test

natecook1000 commented on stdlib/public/core/CString.swift:
```diff
@@ -419,19 +473,32 @@ extension String {
+  ///     interpreted.
+  @inlinable
   @_alwaysEmitIntoClient
+  @available(swift, deprecated: 6, message: "Use String(decoding: array, as: Encoding.self) instead")
```
Out of curiosity, why are these depreciations for Swift 6? Are the replacements not added until then?

glessard replied:
The cleanest way to soft-deprecate is to hang the deprecation on a future language mode; Xcode de-emphasizes the deprecated spelling in suggestions, and docc should indicate the deprecation. Renamings are a mild source-break, so having them kick in with the language mode is a good time. We could also decide to bring the deprecations forward at a later time (e.g. Swift 5.19).

grynspan replied:
swift-testing also needs a bit of runway to adopt, because we need to wait for people to have toolchains with the new function before we can safely write:

```swift
#if swift(>=5.11)
useNewInit()
#else
useOldInit()
#endif
```

(We're planning to adopt once various dusts settle.)

natecook1000 reviewed (commented):
LGTM! Are the specialization removals intended in this PR as well?

glessard commented:
@swift-ci please test

glessard committed:
[stdlib] make parameter label the same as in peer functions

These functions were the odd ones out in not using `encoding` for this parameter label. Harmonizing this term will also make the documentation more consistent.

glessard commented:
@swift-ci please test

glessard commented:
> Are the specialization removals intended in this PR as well?

These were all on `@_alwaysEmitIntoClient` functions, so as far as I can tell they were useless.

grynspan commented on stdlib/public/core/CString.swift:
```diff
@@ -213,9 +258,21 @@ extension String {
+  ///
+  /// Note: This initializer is deprecated. Use the initializer
+  ///       `String.init?(validating: array, as: UTF8.self)` instead,
+  ///       remembering that "\0" is a valid character in Swift.
```
The fact that the trailing null is going to get absorbed into the Swift string makes this _not_ a drop-in replacement. Developers will see the deprecation message, but not the note about null characters. I would suggest instead telling developers to use `array.withUnsafeBuffer {}` so they can use the other `validatingCString:` initializer that takes a pointer.

glessard replied:
If there is a known trailing zero, one can use `String(validating: myTerminatedArray.dropLast(), as: UTF8.self)`, and that's better than using the unsafe hatch. I don't think the note should be overly prescriptive, though.

grynspan replied:
I _can_ do that, sure, but if I'm unfamiliar with the differences in API, I'm presumably just going to do what the diagnostic tells me to (replace this call with one to the other API) and then I'll introduce a correctness bug that I might not even notice at first.

We don't need an overly verbose diagnostic, but the one we emit should provide clear and correct instructions for resolving the issue. :)

grynspan commented on stdlib/public/core/CString.swift:
```diff
@@ -198,8 +230,21 @@ extension String {
+  ///     An array containing a null-terminated sequence of UTF-8 code units.
   @inlinable
   @_alwaysEmitIntoClient
+  @available(swift, deprecated: 6, message: "Use String(validating: array, as: UTF8.self) instead")
```
As with the other initializer, the null character poses a problem. Since this initializer is basically _never_ desireable, is it possible to mark it `unavailable` rather than deprecated, or will that prevent it from participating in overload resolution (despite being an exact match)?

grynspan replied:
Assuming the overload resolution still picks it, anyway.

glessard replied:
Your intuition was correct: overload resolution skips over unavailable or obsoleted overloads. Oh well.

glessard committed:
[doc] fix note visibility for docc-generated pages

glessard committed:
[test] fix diagnostics tests for String-from-C-string inits

glessard committed:
[stdlib] tweak deprecation message

glessard commented:
@swift-ci please test
